### PR TITLE
[COv4] Fix code object v4 support

### PR DESCRIPTION
**Pull request by atamazov**
- Fix for CO v4 version auto-detection introduced in #582.
- Fix for CO-option-syntax auto-detection introduced in #582.
- Introduced `MIOPEN_DEBUG_OPENCL_ENFORCE_CODE_OBJECT_OPTION` (undocumented intentionally)

### CODE REVIEW

File: src/hipoc/hipoc_program.cpp [hunk 1]
```diff
@@ -59,34 +66,73 @@ namespace miopen {
...
+    auto syntax = miopen::Value(MIOPEN_DEBUG_OPENCL_ENFORCE_CODE_OBJECT_OPTION{});
+    if(syntax > 4)
+    {
+        MIOPEN_LOG_E("Bad MIOPEN_DEBUG_OPENCL_ENFORCE_CODE_OBJECT_OPTION, using default");
+        syntax = 0;
+    }
+
+    if(syntax == 0)
+    {
+        if(HipCompilerVersion() >= external_tool_version_t{4, 0, 999999}) /// \ todo
```
Comment (atamazov) [inline src/hipoc/hipoc_program.cpp:80]: This should be updated when the compiler will accept "-mcode-object-version = N".

File: src/hipoc/hipoc_program.cpp [hunk 2]
```diff
@@ -59,34 +66,73 @@ namespace miopen {
...
-        code_object_version = 0;
+        co_version = 0;
     }
 
-    if(HipCompilerVersion() >= external_tool_version_t{4, 0, -1})
+    if(co_version == 0)
     {
-        if(code_object_version == 0) // Env.var is unset.
-            return {};               // Rely on compiler's default.
+        if(HipCompilerVersion() >= external_tool_version_t{4, 0, 999999}) /// \ todo
```
Comment (atamazov) [inline src/hipoc/hipoc_program.cpp:101]: This should be updated when the whole stack (HIP compiler + runtime) supports CO v4.

**Review by atamazov — COMMENTED** [review]
Comments for reviewers.

**Review by daniellowell — APPROVED** [review]

**Comment by atamazov** [comment]
I am terminating it's CI job to save resources.

**Pull request merged**